src/dos/program_autotype.cpp

```diff
@@ -0,0 +1,158 @@
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <string>
+#include <sstream>
+
+#include "support.h"
+#include "mapper.h"
+#include "dosbox.h"
+#include "programs.h"
+#include "program_autotype.h"
+
+void AUTOTYPE::PrintUsage() {
+	WriteOut(
+		"\033[32;1mAUTOTYPE\033[0m [-list] [-w WAIT] [-p PACE] button_1 [button_2 [...]] \n\n"
+		"Where:\n"
+		"  -list:   prints all available button names.\n"
+		"  -w WAIT: seconds before typing begins. Two second default; max of 30.\n"
+		"  -p PACE: seconds between each keystroke. Half-second default; max of 10.\n"
+		"\n"
+		"  The sequence is comprised of one or more space-separated buttons.\n"
+		"  Autotyping begins after WAIT seconds, and each button is entered \n"
+		"  every PACE seconds. The , character inserts an extra PACE delay.\n"
+		"\n"
+		"Some examples:\n"
+		"  \033[32;1mAUTOTYPE\033[0m -w 1 -p 0.3 up enter , right enter\n"
+		"  \033[32;1mAUTOTYPE\033[0m -p 0.2 1 3 , , enter\n"
+		"  \033[32;1mAUTOTYPE\033[0m -w 1.3 esc esc esc enter p l a y e r enter\n");
+}
+
+// Prints the key-names for the mapper's currently-bound events.
+void AUTOTYPE::PrintKeys() {
+		const std::vector<std::string> names = MAPPER_GetEventNames("key_");
+
+		// Keep track of the longest key name
+		size_t max_length = 0;
+		for (const auto &name : names)
+			max_length = (std::max)(name.length(), max_length);
+
+		// Sanity check to avoid dividing by 0
+		if (!max_length) {
+			WriteOut("AUTOTYPE: The mapper has no key bindings\n");
```

Comment:
For simple strings there's `WriteOut_NoParsing` (or some silly similar name, I don't remember exactly), which won't route the string via `s*printf`.

Reply:
Thanks! Dropped this.